src/main/java/org/javarosa/core/util/externalizable/ExtWrapList.java

```diff
@@ -3,66 +3,83 @@
 import java.io.DataInputStream;
 import java.io.DataOutputStream;
 import java.io.IOException;
+import java.util.List;
 import java.util.Vector;
 
 
-//list of objects of single (non-polymorphic) type
+// List of objects of single (non-polymorphic) type
 public class ExtWrapList extends ExternalizableWrapper {
     public ExternalizableWrapper type;
     private boolean sealed;
+    private Class<? extends List> listImplementation;
 
-    /* serialization */
+    /* Constructors for serialization */
 
-    public ExtWrapList(Vector val) {
+    public ExtWrapList(List val) {
         this(val, null);
     }
 
-    public ExtWrapList(Vector val, ExternalizableWrapper type) {
+    public ExtWrapList(List val, ExternalizableWrapper type) {
         if (val == null) {
             throw new NullPointerException();
         }
 
         this.val = val;
         this.type = type;
+        this.listImplementation = val.getClass();
     }
 
-    /* deserialization */
+    /* Constructors for deserialization */
 
     public ExtWrapList() {
 
     }
 
-    public ExtWrapList(Class type) {
-        this(type, false);
+    // Assumes that the list implementation is a Vector, since that is most common
+    public ExtWrapList(Class listElementType) {
+        this(listElementType, Vector.class);
     }
 
-    public ExtWrapList(Class type, boolean sealed) {
-        this.type = new ExtWrapBase(type);
-        this.sealed = sealed;
+    public ExtWrapList(Class listElementType, Class listImplementation) {
+        this.type = new ExtWrapBase(listElementType);
+        this.listImplementation = listImplementation;
+        this.sealed = false;
     }
 
+    // Assumes that the list implementation is a Vector, since that is most common
     public ExtWrapList(ExternalizableWrapper type) {
+        this(type, Vector.class);
+    }
+
+    public ExtWrapList(ExternalizableWrapper type, Class listImplementation) {
         if (type == null) {
             throw new NullPointerException();
         }
 
+        this.listImplementation = listImplementation;
         this.type = type;
     }
 
     @Override
     public ExternalizableWrapper clone(Object val) {
-        return new ExtWrapList((Vector)val, type);
+        return new ExtWrapList((List)val, type);
     }
 
     @Override
     public void readExternal(DataInputStream in, PrototypeFactory pf) throws IOException, DeserializationException {
         if (!sealed) {
             int size = (int)ExtUtil.readNumeric(in);
-            Vector<Object> v = new Vector<>(size);
-            for (int i = 0; i < size; i++) {
-                v.addElement(ExtUtil.read(in, type, pf));
+            try {
+                List<Object> l = listImplementation.newInstance();
```

Comment:
Initializing the list/vector with its size is actually a pretty serious performance win in the serialization layer, because it prevents alloc/gc's from needing to occur. 

It would even be fine (just for preventing performance regressions) to check if listImplemenation is a vector, and if so allocating it with new Vector(size), and using argumentless newInstance() otherwise. That won't let the other types of lists have that performance gain, but will prevent existing code from accidentally running slower, since they'll all be Vectors under the hood for now anyway.

Reply:
Ah cool makes sense, thanks for catching